data/src/main/scala/cats/data/IList.scala

```diff
@@ -0,0 +1,276 @@
+  final def show(implicit A: Show[A]): String =
+    "[" + toNel.fold("")(_.reduceLeft(A.show)(_ + "," + A.show(_))) + "]"
+
+  /** compute the size of an [[IList]] */
+  final def size: Int = {
+    @inline @tailrec def loop(as: IList[A], acc: Int): Int = as match {
+      case ICons(_, t) => loop(t, acc + 1)
+      case INil()      => acc
+    }
+    loop(this, 0)
+  }
+
+  /** get the tail if the [[IList]] is not empty */
+  final def tailOption: Option[IList[A]] = this match {
+    case INil()      => Option.empty
+    case ICons(_, t) => Some(t)
+  }
+
+  /** take the `n` first elements */
+  final def take(n: Int): IList[A] = {
+    @tailrec
+    def loop(as: IList[A], m: Int, acc: IList[A]): IList[A] = as match {
+      case INil() => acc.reverse
+      case ICons(h,t) => if(m > 0) loop(t, m - 1, ICons(h, acc)) else acc.reverse
+    }
+    loop(this, n, empty)
+  }
+
+  /** take elements as long as the predicate holds */
+  final def takeWhile(p: A => Boolean): IList[A] = {
+    @tailrec
+    def loop(as: IList[A], acc: IList[A]): IList[A] = as match {
+      case INil() => acc.reverse
+      case ICons(h,t) => if(p(h)) loop(t, ICons(h, acc)) else acc.reverse
+    }
+    loop(this, empty)
+  }
+
+  /** transform an [[IList]] into a [[scala.List]] */
+  final def toList: List[A] =
+    reverse.foldLeft(List.empty[A])((acc, a) => a :: acc)
+
+  /** attempt to transform an [[IList]] into a [[NonEmptyList]] */
+  final def toNel: Option[NonEmptyList[A]] =
+    uncons.map{ case (h,t) => NonEmptyList(h, t) }
+
+  /**
+   * `toString` will give the same result as `show` if there exists a `Show[A]`
+   * instance and `Show[A].show` is equal to `toString`
+   */
+  override final def toString: String =
+    "[" + toNel.fold("")(_.reduceLeft(_.toString)((acc, a) => a.toString + "," + acc)) + "]"
+
+  final def traverse[G[_], B](f: A => G[B])(implicit G: Applicative[G]): G[IList[B]] =
+    reverse.foldLeft(G.pure(empty[B]))((acc, a) => G.map2(acc, f(a))(_.prepend(_)))
+
+  /** attempt to get head and tail of an [[IList]] */
+  final def uncons: Option[(A, IList[A])] = this match {
+    case INil()      => Option.empty
+    case ICons(h, t) => Some((h,t))
+  }
+
+  /** widen the type of an [[IList]] */
+  final def widen[B >: A]: IList[B] =
+    this.asInstanceOf[IList[B]]
+
+  /** check if two matches are equal */
+  final def ===(other: IList[A])(implicit A: Eq[A]): Boolean = {
+    @inline @tailrec
+    def loop(as: IList[A], bs: IList[A]): Boolean = (as, bs) match {
+      case (INil(), INil()) => true
+      case (ICons(x, xs), ICons(y, ys)) => A.eqv(x,y) && loop(xs, ys)
+      case _ => false
+    }
+    loop(this, other)
+  }
+
+}
+
+object IList extends IListInstances {
+  final case class INil[A]() extends IList[A]
+  final case class ICons[A](head: A, tail: IList[A]) extends IList[A]
+
+  private val nil: IList[Nothing] = INil()
+
+  /** create an [[IList]] with a single element */
+  def singleton[A](a: A): IList[A] =
+    ICons(a, INil())
```

Comment:
Could use `empty[A]` here, since it's used in other methods like `filter`, `flatMap`, etc.